src/runtime/contrib/tensorrt/tensorrt_runtime.cc

```diff
@@ -174,25 +176,50 @@ class TensorRTRuntime : public JSONRuntimeBase {
       int binding_index = engine->getBindingIndex(name.c_str());
       ICHECK_NE(binding_index, -1);
       if (data_entry_[eid]->device.device_type != kDLCUDA) {
-        device_buffers[binding_index].CopyTo(const_cast<DLTensor*>(data_entry_[eid]));
+        auto device_buffer = GetOrAllocateDeviceBuffer(eid, binding_index);
+        device_buffer.CopyTo(const_cast<DLTensor*>(data_entry_[eid]));
       }
     }
   }
 
  private:
+  /*! \brief Get batch size for engine from the runtime input shapes. */
+  int GetBatchSize() {
+    return data_entry_[input_var_eid_[0]]->ndim == 0 ? 1 : data_entry_[input_var_eid_[0]]->shape[0];
+  }
+
+  /*! \brief TensorRT engines are built for a maximum batch size. If an engine doesn't exist for a
+   * certain batch size already, see if we can reuse an engine built for a higher batch size. */
+  bool FindCompatibleEngine(int batch_size, int* compatible_engine_batch_size) {
+    // Check for exact match
+    if (trt_engine_cache_.count(std::make_pair(symbol_name_, batch_size))) {
+      *compatible_engine_batch_size = batch_size;
+      return true;
+    }
```

Comment:
1. IIUC, seems like you don't need the exact match, so we can rely on the following logic and remove this one. This can also reduce the number of cached engines.
2. Accordingly, can we only keep one engine with the so far largest batch size? i.e., after we build a new engine with larger batch size, can we throw away the old one?

Reply:
Thanks for the review! Yes, I suppose we could do that instead.

Reply:
From TRT's documentation, it sounds like doing this could have performance impact.

"Another consideration is that building the optimized network optimizes for the given maximum batch size. The final result will be tuned for the maximum batch size but will still work correctly for any smaller batch size. It is possible to run multiple build operations to create multiple optimized engines for different batch sizes, then choose which engine to use based on the actual batch size at runtime. "


The implementation in this PR isn't much better because it depends on the order of batch sizes encountered. For example, if we encounter batch sizes 1, 2, 4 then 3 engines are built. But if we encounter in the order 4, 2, 1, only one engine is built and it will be suboptimal for input sizes 1 and 2 which is pretty much the same result as if we only kept one engine with the largest batch size.

Reply:
Exactly, and this is a common issue also happening at TVM tuning. I'm not sure how TRT users would prefer to build the best fit engine every time or willing to reuse the built one with some performance loss. Maybe making it configurable first might be better.

Reply:
I see. Having it configurable makes the most sense to me also. We can have two modes, "best fit" which was the previous behavior, or "single engine" which would be built for the highest batch size. I will update the PR when done.

Reply:
Added env variable `TVM_TENSORRT_MULTI_ENGINE` to configure this.